### PR TITLE
#63 improved logging of ansible tasks

### LINKED ISSUE #61: Change initial password of jupyter notebook
Currently: "script-languages"
Expected: "dss"

```[tasklist]
### Tasks
- [x] update user guide
-  [x] update file `exasol/ds/sandbox/runtime/ansible/roles/jupyter/defaults/main.yml`
```
Content of ansible task: 
```
 password: "{{ lookup('ansible.builtin.env', 'JUPYTER_LAB_PASSWORD', default='script-languages') }}"
```

### LINKED ISSUE #63: Improve logging of Ansible tasks
Currently in case of a failure the ci-build does not show the original cause, e.g. outdated version of an ubuntu package pinned in the ansible scripts. 
Example: https://github.com/exasol/data-science-sandbox/actions/runs/6877900023/job/18706405162

Additionally in support for investigations requested by #59 this ticket requests to report the  duration of single ansible steps.

### PR DESCRIPTION
**Pull request by ckunki**
Closes #63 
closes #61

### CODE REVIEW

File: test/integration/test_create_dss_docker_image.py [hunk 1]
```diff
@@ -16,6 +18,22 @@
 from exasol.ds.sandbox.lib import pretty_print
 
 
+class StderrCapture:
```
Comment (ckunki) [inline test/integration/test_create_dss_docker_image.py:21]: Currently unused, see [comment](https://github.com/exasol/data-science-sandbox/issues/59#issuecomment-1814510146) in ticket #59.

File: exasol/ds/sandbox/runtime/ansible/general_setup_tasks.yml [hunk 2]
```diff
@@ -3,6 +3,9 @@
     upgrade: yes
     update_cache: yes
   become: "{{need_sudo}}"
+- name: Install rsync
+  include_role:
+    name: rsync
```
Comment (ckunki) [inline exasol/ds/sandbox/runtime/ansible/general_setup_tasks.yml:8]: We could also skip installation of `rsync` for now and revert `tutorial.yml` to use `copy` again, as installation of `rsync` itself takes quite long and hence in total there is no significant benefit over `ansible.builtin.copy`.

File: exasol/ds/sandbox/lib/ansible/ansible_access.py [hunk 3]
```diff
@@ -11,19 +13,32 @@ class AnsibleException(RuntimeError):
...
     """
     Provides access to ansible runner.
     @raises: AnsibleException if ansible execution fails
     """
     @staticmethod
-    def run(private_data_dir: str, run_ctx: AnsibleRunContext, printer: Callable[[str], None]):
+    def run(
+            private_data_dir: str,
+            run_ctx: AnsibleRunContext,
+            printer: Callable[[str], None],
```
Comment (Nicoretti) [inline exasol/ds/sandbox/lib/ansible/ansible_access.py:26]:
```suggestion
            logger: Callable[[str], None],
```

File: exasol/ds/sandbox/lib/ansible/ansible_access.py [hunk 4]
```diff
@@ -11,19 +13,32 @@ class AnsibleException(RuntimeError):
...
+        r = ansible_runner.run(
+            private_data_dir=private_data_dir,
+            playbook=run_ctx.playbook,
+            quiet=quiet,
+            event_handler=event_handler,
+            extravars=run_ctx.extra_vars,
+        )
         for e in r.events:
-            printer(e)
+            printer(json.dumps(e, indent=2))
```
Comment (Nicoretti) [inline exasol/ds/sandbox/lib/ansible/ansible_access.py:38]:
```suggestion
            logger.debug(json.dumps(e, indent=2))
```

File: exasol/ds/sandbox/lib/ansible/ansible_access.py [hunk 5]
```diff
@@ -1,7 +1,9 @@
 import ansible_runner
```
Comment (Nicoretti) [inline exasol/ds/sandbox/lib/ansible/ansible_access.py:1]: Having a module `ansible_access` which just contains a class `AnsibleAccess` does not make much sense and only increases noise and verbosity imho. (`ansible_access.AnsibleAcces` vs `ansible.Access` )

`ansible.Access` should be as descriptive.

File: exasol/ds/sandbox/lib/ansible/ansible_runner.py [hunk 6]
```diff
@@ -22,8 +23,24 @@ def __init__(self, ansible_access: AnsibleAccess, work_dir: Path):
     def printer(msg: str):
         LOG.debug(msg)
```
Comment (Nicoretti) [inline exasol/ds/sandbox/lib/ansible/ansible_runner.py:24]: not needed just pass `LOG.debug` as printer

File: exasol/ds/sandbox/lib/ansible/ansible_runner.py [hunk 7]
```diff
@@ -22,8 +23,24 @@ def __init__(self, ansible_access: AnsibleAccess, work_dir: Path):
...
 
+    @staticmethod
+    def event_handler(event: Dict[str, any]) -> bool:
+        try:
+            duration = event["event_data"]["duration"]
+            if duration is not None and duration > 0.5:
+                print(f"duration: {round(duration)} seconds")
+        except KeyError as ex:
```
Comment (Nicoretti) [inline exasol/ds/sandbox/lib/ansible/ansible_runner.py:32]: ?

File: exasol/ds/sandbox/lib/ansible/ansible_runner.py [hunk 8]
```diff
@@ -22,8 +23,24 @@ def __init__(self, ansible_access: AnsibleAccess, work_dir: Path):
...
     def run(self, ansible_run_context: AnsibleRunContext, host_infos: Tuple[HostInfo]):
         inventory_content = render_template("inventory.jinja", host_infos=host_infos)
         with open(self._work_dir / "inventory", "w") as f:
             f.write(inventory_content)
-        self._ansible_access.run(str(self._work_dir), ansible_run_context, self.printer)
+        event_handler = self.event_handler if LOG.isEnabledFor(logging.INFO) else None
+        self._ansible_access.run(
+            str(self._work_dir),
+            ansible_run_context,
+            self.printer,
```
Comment (Nicoretti) [inline exasol/ds/sandbox/lib/ansible/ansible_runner.py:44]:
```suggestion
            LOG.debug,
```

File: test/integration/test_create_dss_docker_image.py [hunk 9]
```diff
@@ -16,6 +18,22 @@
...
+        sys.stdout = self._buffer
+
+    def stop(self):
+        sys.stdout = self._stream
+
+    @property
+    def output(self) -> str:
+        """Return the captured output."""
+        self._buffer.seek(0)
+        return self._buffer.read().strip()
```
Comment (Nicoretti) [inline test/integration/test_create_dss_docker_image.py:34]: where is `sys.stderr`? I only see `sys.stdout` the class name states it is `StderrCapture` though.
The context lib may be of help here [redirect_stdout](https://docs.python.org/3/library/contextlib.html?highlight=context%20manager#contextlib.redirect_stdout)

File: exasol/ds/sandbox/lib/ansible/ansible_access.py (hunk 5 above)
Comment (ckunki) [reply exasol/ds/sandbox/lib/ansible/ansible_access.py:1]: You're speaking from my soul, Nico.
See my (removed) comment in [commit 5e66b3405](https://github.com/exasol/data-science-sandbox/pull/40/commits/5e66b34051f9783f3f41c4985762959a2720fab6#diff-6870d7bc222dcc2be29df4ec484c7bd1ca0b666a7ace84497069a3b57c3ddcde).

Could we address the renaming in the scope of ticket #65 and I'd love to kick it off in a pair-discussion?

File: exasol/ds/sandbox/runtime/ansible/general_setup_tasks.yml (hunk 2 above)
Comment (tkilias) [reply exasol/ds/sandbox/runtime/ansible/general_setup_tasks.yml:8]: We probably will get more files in the long-term, so rsync might amortize better then copy

File: test/integration/test_create_dss_docker_image.py [hunk 10]
```diff
@@ -16,6 +18,22 @@
...
```
Comment (Nicoretti) [inline test/integration/test_create_dss_docker_image.py:34]: delete?

File: exasol/ds/sandbox/lib/ansible/ansible_runner.py [hunk 11]
```diff
@@ -17,13 +19,36 @@ class AnsibleRunner:
...
+        if duration > 0.5:
+            self._duration_logger.debug(f"duration: {round(duration)} seconds")
+        return True
 
     def run(self, ansible_run_context: AnsibleRunContext, host_infos: Tuple[HostInfo]):
         inventory_content = render_template("inventory.jinja", host_infos=host_infos)
         with open(self._work_dir / "inventory", "w") as f:
             f.write(inventory_content)
-        self._ansible_access.run(str(self._work_dir), ansible_run_context, self.printer)
+        event_handler = self.event_handler if LOG.isEnabledFor(logging.INFO) else None
```
Comment (Nicoretti) [inline exasol/ds/sandbox/lib/ansible/ansible_runner.py:48]:
```suggestion
        event_handler = self.event_handler if LOG.isEnabledFor(logging.INFO) else lambda msg: None
```

**Review by Nicoretti — APPROVED** [review]

**Pull request merged**